Source/FieldSolver/ElectrostaticSolver.cpp

```diff
@@ -346,17 +370,126 @@ WarpX::computePhiCartesian (const amrex::Vector<std::unique_ptr<amrex::MultiFab>
 
     // Solve the Poisson equation
     linop.setDomainBC( lobc, hibc );
+
+    for (int lev=0; lev < rho.size(); lev++){
+        rho[lev]->mult(-1._rt/PhysConst::ep0);
+    }
+
     MLMG mlmg(linop);
     mlmg.setVerbose(2);
     mlmg.setMaxIter(max_iters);
     mlmg.solve( GetVecOfPtrs(phi), GetVecOfConstPtrs(rho), required_precision, 0.0);
+}
+#endif
 
-    // Normalize by the correct physical constant
-    for (int lev=0; lev < rho.size(); lev++){
-        phi[lev]->mult(-1._rt/PhysConst::ep0);
+/* \bried Set Dirichlet boundary conditions for the electrostatic solver.
+
+    The given potential's values are fixed on the boundaries of the given
+    dimension according to the desired values from the simulation input file,
+    boundary.potential_lo and boundary.potential_hi.
+
+   \param[inout] phi The electrostatic potential
+   \param[in] idim The dimension for which the Dirichlet boundary condition is set
+*/
+void
+WarpX::setPhiBC( amrex::Vector<std::unique_ptr<amrex::MultiFab> >& phi,
+                 std::array<bool,AMREX_SPACEDIM> dirichlet_flag,
+                 Array<amrex::Real,AMREX_SPACEDIM> phi_bc_values_lo,
+                 Array<amrex::Real,AMREX_SPACEDIM> phi_bc_values_hi ) const
+{
+    // check if any dimension has Dirichlet boundary conditions
+    bool has_Dirichlet = false;
+    for (int idim=0; idim<AMREX_SPACEDIM; idim++){
+        if (dirichlet_flag[idim]) {
+            has_Dirichlet = true;
+        }
     }
-}
+    if (!has_Dirichlet) return;
+
+    amrex::Box domain = Geom(0).Domain();
+    domain.surroundingNodes();
+    const auto lo = lbound(domain);
+    const auto hi = ubound(domain);
+
+    // loop over all mesh refinement levels and set the boundary values
+    for (int lev=0; lev <= max_level; lev++) {
+
+#ifdef AMREX_USE_OMP
+#pragma omp parallel if (amrex::Gpu::notInLaunchRegion())
 #endif
+        for ( MFIter mfi(*phi[lev], TilingIfNotGPU()); mfi.isValid(); ++mfi ) {
+            // Extract the potential
+            auto phi_arr   = phi[lev]->array(mfi);
+            // Extract tileboxes for which to loop
+            const Box& tb  = mfi.tilebox( phi[lev]->ixType().toIntVect() );
+
+            amrex::ParallelFor( tb,
+                [=] AMREX_GPU_DEVICE (int i, int j, int k) {
+                    int idx, idx_lo, idx_hi;
+
+                    for (int idim=0; idim<AMREX_SPACEDIM; idim++){
```

Comment:
Can the loop over idim (and the check of dirichlet_flag) be moved outside of the ParallelFor loop? This would allow skipping the entire ParallelFor when the dirichlet_flag for a dimension is false. The setting of idx_lo and hi can then also be moved out of the ParallelFor. This would help in addressing my comment about avoiding the overhead when the operation is not needed.

Reply:
Thanks for the comments! I agree the unnecessary overhead is not good. I can add a check to compare the boundary condition from the previous step with the desired condition for the current step and if those two are the same I'll return early for that dimension. Also happy to move the checks you mentioned outside the ParallelFor.

Reply:
Considering this more, moving the loops over idim outside the ParallelFor may not be an improvement. On the GPU, there is overhead associated with each ParallelFor to start up the GPU threads. That overhead may overwhelm the overhead of the extra work. Not sure what is the best approach. Perhaps using the ParallelFor over multiple boxes, `ParallelFor(tb1, tb2, tb3, ...)`?

Reply:
I just noticed this comment after pushing the changes as discussed earlier. I don't really have any insight into the relative overhead between the two approaches. Our use case is primarily simulations with periodic boundary conditions in all directions except 1, we usually set Dirichlet BCs in the z-direction. So in that case either approach will result in just one ParallelFor being spawned. How would the ParallelFor over multiple boxes approach work?

The commit I just pushed also included switching to `boundary.field_lo/hi`. On the python side I changed the argument to match the format for specifying a 'field' boundary conditions through `warpx_bc_xmax_field`, for example. Hopefully this keeps things clear. On internal variables I've opted for continuing to refer to it as a potential. I didn't add a new option for `dirichlet` to `FieldBCType_algo_to_int`, since the purpose of the Dirichlet BC is to simulate a perfect conductor at the walls with a fixed potential applied to it, so it should be the same as the PEC boundary condition, or am I missing a distinction?

Reply:
It is a good point that common use cases will not be setting the potential on all boundaries, so having the loops as they are now is probably ok. I would leave it as is now, but maybe add a comment to the code saying that testing is needed to verify that this structure is optimal.

Regarding the input, I would keep both the field_lo,hi and potential_lo,hi. The field input would be the type of BC, i.e. PEC, and the potential would be the values. Also, since the potential typically won't be specified for all boundaries, perhaps separate the input, boundary.potential_lo_x etc.

Reply:
Sounds good, I'll add the comment.

Oh, I see I misunderstood regarding the `field_lo` point. So just to be explicit you'd like me to change the check in `ElectrostaticSolver.cpp`, `if ( Geom(0).isPeriodic(idim) )` to 
```
    if ( WarpX::field_boundary_lo[1] == FieldBoundaryType::Periodic
         || WarpX::field_boundary_hi[1] == FieldBoundaryType::Periodic ) {
```
I guess we currently just assume that anything other than periodic would result in Dirichlet BCs for the electrostatic solver, do you think I should change this to only accept the PEC condition for a Dirichlet BC solve?

I'm thinking it would also be a good change to build the parsers (currently done in `getPhiBC`) in `WarpXUtil.cpp::ReadBCParams()`.

Reply:
Yes, that change would be good, to check if the lo and hi are periodic. Since the BC types are now explicitly specified, I would check that the lo and hi are set to either periodic or PEC and otherwise raise an error and exit (using an `AMREX_ALWAYS_ASSERT_WITH_MESSAGE`). This would help avoid user errors.

For building the parsers, I would keep it as it is currently done in `getPhiBC`. `ReadBCParams` is where it is since it is doing setup that is needed before the geometry input is handled in AMReX code. That is not needed when reading in the potential expressions.

Reply:
Sounds good on the first point.
For the second, do you think having to parse the input and build the parsers at every step is creating significant overhead or are those actions sufficiently fast that we don't mind doing it on every step?

Reply:
I think building the parser would be insignificant time, but there might be overhead if it needs to be transferred to the GPU. I would leave it as it as and worry about such optimizations later. Subsequent PRs can always be added later for fixes.

Reply:
I believe I included everything we discussed above in the latest commit. Besides the change to use `FieldBoundaryType`, I changed the expected input to be in the form `boundary.potential_lo_x` which also allowed me to remove a default potential from the python side (now the default is only specified in C++ code, in `getPhiBC`).